ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -155,6 +156,48 @@ impl Blockstore {
         to_slot: Slot,
         purge_type: PurgeType,
         purge_stats: &mut PurgeStats,
+    ) -> Result<bool> {
+        self.run_purge_with_stats_and_maybe_cleanup_slot_meta(
+            from_slot,
+            to_slot,
+            purge_type,
+            purge_stats,
+            None,
+        )
+    }
+
+    pub(crate) fn run_purge_and_cleanup_slot_meta(
+        &self,
+        slot: Slot,
+        slot_meta: SlotMeta,
+    ) -> Result<bool> {
+        self.run_purge_with_stats_and_maybe_cleanup_slot_meta(
+            slot,
+            slot,
+            PurgeType::Exact,
+            &mut PurgeStats::default(),
+            Some(slot_meta),
+        )
+    }
+
+    /// A helper function to `purge_slots` that executes the ledger clean up.
+    /// The cleanup applies to \[`from_slot`, `to_slot`\].
+    ///
+    /// When `from_slot` is 0, any sst-file with a key-range completely older
+    /// than `to_slot` will also be deleted.
+    ///
+    /// If `slot_meta` is specified for some `child_slot`, we require that
+    /// `child_slot == from_slot == to_slot` - we are purging only one slot.
+    /// In this case along with the purge we remove `child_slot` from its
+    /// `parent_slot_meta.next_slots` as well as reinsert an orphaned `slot_meta`
+    /// for `child_slot` that only retains the `next_slots` value.
+    pub(crate) fn run_purge_with_stats_and_maybe_cleanup_slot_meta(
```

Comment:
Instead of another function / more arguments, how about adding a new case to the `PurgeType`. Maybe something like `ExactAndCleanupChaining`. We already have a spot where we switch on `purge_type` in `Blockstore::run_purge_with_stats()` so I think this would tie in nicely.

Reply:
Actually, I don't think we need a separate type; I think we should just update the logic in the `Exact` case. The non-test case callers that use `PurgeType::Exact` are:
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/core/src/validator.rs#L2189-L2190
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger/src/blockstore.rs#L1177-L1178
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger/src/blockstore/blockstore_purge.rs#L84-L85
    - This code path only reached from https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger-tool/src/blockstore.rs#L942-L948
    
The case we wanted to update is the second bullet, and we absolutely want to fixup chaining for that case. The other two cases are both accompanied by calls to `Blockstore::purge_from_next_slots()` which is essentially doing the same logic as `Blockstore::clear_unconfirmed_slot()` in cleaning up the chaining.

So, It seems we can condense more code

Reply:
`purge_from_next_slots()` looks like it would have pretty terrible performance: https://github.com/solana-labs/solana/blob/ba189b14542f96d63a9b2f7a2588f49d65c74f12/ledger/src/blockstore/blockstore_purge.rs#L98-L102. It starts iterating from slot 0 👻

Reply:
Yeah, the function definitely needs some work. Consider the following graph:
```
4 - 5 - 6 - 9 - x
 \
  7 - 8 - x
```  
If we call with `from_slot = 6, to_slot = 9`, we'd expect the following afterwards:
```
4 - 5 - x
```
But, if we start with `from_slot`'s (`6`) parent (`5`), `4` would still have `7` in its' `next_slots` despite `7` having been purged.

So, I agree that we can iterate `[from_slot, to_slot]` instead of `[0, to_slot]`. But. for every slot in that range, I think we need to update the parent's slot, similar to what `clear_unconfirmed_slot()` is currently doing. So, overhead will look pretty similar for the `clear_unconfirmed_slot()` case where `from_slot == to_slot`.

Reply:
Another thing that just occurred to me - by moving the functionality within the same write batch, we should ensure that the ordering in write batch is respected.

Suppose we have this chain:
```
4 - 5 - 6
```
If we call with `from_slot = 5, to_slot = 5`, there should still be an empty `SlotMeta` for `5` that has `6` in its' `next_slots` vector.

But, earlier in the function, we will have range deleted `[5, 5]`. So, we need to ensure that the `SlotMeta` update comes after the range delete AND double check that ordering is ensured (ie the `SlotMeta` for `5` still present when the function finishes. I believe this to be true, but again, just calling it out since I thought of it.

The logic here seems tricky enough that a unit test might be in order too ... sorry Ashwin, no good deed goes unpunished. Maybe we can whiteboard this out in the next week to make sure we cover all the edges